### PR TITLE
Improve search for FDB cluster files

### PR DESCRIPTION
**Pull request by kocolosk**
## Overview

This patch introduces a handful of related improvements to the code that chooses the FoundationDB cluster file. We now check for a cluster file in up to 4 locations, in priority order:

1. The value of `[erlfdb] cluster_file` in the server config
2. The value of the FDB_CLUSTER_FILE environment variable
3. /usr/local/etc/foundationdb/fdb.cluster (MacOS package default)
4. /etc/foundationdb/fdb.cluster (Linux packages default)

The second and fourth locations are new. We also check that the file has appropriate access permissions, namely that is RW accessible by the user running CouchDB. Finally, we add some more detailed logging to help users debug any problems in this department on startup.

If a location is specified using either one of the first two "custom" approaches and there is a problem with that location, CouchDB will abort rather than try any further option. This avoids a situation where a user intended to override the default cluster file but CouchDB starts up with the default one because e.g. the user failed to actually install the new file with the correct permissions.

## Testing recommendations

Still need to add a test suite here, and really some documentation too (although that would probably be a larger project involving documentation of the 4.0 stack as a whole). I could use a bit of guidance on the testing. Should I just verify the behavior of `find_cluster_file/1` with a mix of valid and invalid input? Is it better to create fixtures with bad permissions or just mock `file:read_file_info/2`?

I've tested several aspects of this manually:
- [x] checked that a custom cluster file with bad permissions causes a crash
- [x] checked that FDB_CLUSTER_FILE can be used instead of the server config setting
- [x] checked that CouchDB will find /etc/foundationdb/fdb.cluster without any config help

## Checklist

- [x] Code is written and works correctly
- [ ] Changes are covered by tests
- [ ] Any new configurable parameters are documented in `rel/overlay/etc/default.ini`
- [ ] A PR for documentation changes has been made in https://github.com/apache/couchdb-documentation

### CODE REVIEW

File: src/fabric/src/fabric2_server.erl [hunk 1]
```diff
@@ -42,9 +42,10 @@
 
 
 -include_lib("couch/include/couch_db.hrl").
+-include_lib("kernel/include/file.hrl").
 
-
--define(CLUSTER_FILE, "/usr/local/etc/foundationdb/fdb.cluster").
+-define(CLUSTER_FILE_MACOS, "/usr/local/etc/foundationdb/fdb.cluster").
+-define(CLUSTER_FILE_LINUX, "/etc/foundationdb/fdb.cluster").
```
Comment (nickva) [inline src/fabric/src/fabric2_server.erl:48]: Wonder if we could add Windows to the list. Not sure how supported it is, but I found the default cluster file location in the Rust library https://docs.rs/foundationdb/0.5.0/src/foundationdb/lib.rs.html#171-174

File: src/fabric/src/fabric2_server.erl [hunk 2]
```diff
@@ -212,14 +213,47 @@ get_db_and_cluster(EunitDbOpts) ->
...
             {ok, ConnectionStr} = file:read_file(ClusterFileStr),
             DbHandle = erlfdb:open(iolist_to_binary(ClusterFileStr)),
             {string:trim(ConnectionStr), DbHandle}
     end,
     apply_tx_options(Db, config:get(?TX_OPTIONS_SECTION)),
     {Cluster, Db}.
 
+get_cluster_file_string() ->
+    Locations = [
+        {custom, config:get("erlfdb", "cluster_file", false)},
```
Comment (nickva) [inline src/fabric/src/fabric2_server.erl:226]: As a minor nit, I wonder if we could switch to use `undefined` since in more cases in our code we rely on `config:get(Section, Key)` returning `undefined` but are not as familiar with `os:getenv/1`. In order to use `undefined` we could use [`os:getenv/2`](https://erlang.org/doc/man/os.html#getenv-2) as in `os:getenv("FDB_CLUSTER_FILE", undefined)` then handle the `undefined` in `find_cluster_file` clauses.

File: src/fabric/src/fabric2_server.erl [hunk 3]
```diff
@@ -212,14 +213,47 @@ get_db_and_cluster(EunitDbOpts) ->
...
+            erlang:error(Reason)
+    end.
+
+find_cluster_file([]) ->
+    {error, cluster_file_missing};
+find_cluster_file([{custom, false} | Rest]) ->
+    find_cluster_file(Rest);
+find_cluster_file([{Type, Location} | Rest]) ->
+    case file:read_file_info(Location, [posix]) of
+        {ok, #file_info{access = read_write}} ->
```
Comment (nickva) [inline src/fabric/src/fabric2_server.erl:244]: This might break some development setups where FDB server system packages are installed (deb, rpm, etc) and run as `root ` (or `foundationdb` user), and CouchDB instance is run as the logged in user. One way to address that edge case is perhaps having the `./dev/run` script make a local writable copy somewhere and use that when starting CouchDB, or may be just have a config setting set by `./dev/run` to skip the `writable` check.

File: src/fabric/src/fabric2_server.erl [hunk 4]
```diff
@@ -212,14 +213,47 @@ get_db_and_cluster(EunitDbOpts) ->
...
+            couch_log:info("Using ~s FDB cluster file: ~s", [Type, Location]),
+            {ok, Location};
+        {ok, _} ->
+            couch_log:error("CouchDB needs read/write access to FDB cluster file: ~s", [Location]),
+            {error, cluster_file_permissions};
+        {error, Reason} when Type =:= custom ->
+            couch_log:error("Encountered ~p error looking for FDB cluster file: ~s", [Reason, Location]),
+            {error, Reason};
+        {error, Reason} when Type =:= default ->
```
Comment (nickva) [inline src/fabric/src/fabric2_server.erl:253]: On Linux hosts this would emit an `Encountered enoent error looking for FDB cluster file: /usr/local/etc/foundationdb/fdb.cluster` warning. Maybe we could add an explicit `enoent` clause to issue an info log message instead of the `enoent` error, since it's an expected scenario.

File: src/fabric/src/fabric2_server.erl [hunk 5]
```diff
@@ -212,14 +213,47 @@ get_db_and_cluster(EunitDbOpts) ->
...
+    case find_cluster_file(Locations) of
+        {ok, Location} ->
+            Location;
+        {error, Reason} ->
+            erlang:error(Reason)
+    end.
+
+find_cluster_file([]) ->
+    {error, cluster_file_missing};
+find_cluster_file([{custom, false} | Rest]) ->
```
Comment (nickva) [inline src/fabric/src/fabric2_server.erl:240]: Minor style nit: for consistency with the rest of the file, let's use double lines between functions and single lines between clauses.

**Review by nickva — COMMENTED** [review]
A nice improvement, both ergonomic and correctness-wise with the writable permission checking.

Tested on Ubuntu 20.04 with system-wide installed fdb server, running as root. Encountered the issue about it picking up the default `/etc/foundationdb/fdb.cluster` file and failing on a writable check. It's probably not worth worrying about that corner case as much for development since a writable copy of the file and `FDB_CLUSTER_FILE` set in `~/.bashrc` seems to work well.

For tests, mocking `file` might be the best option. `file` and `os` modules are `sticky` modules as far as meck is concerned, so would have to do something like `meck:new(file, [unstick, passthrough]).`

File: src/fabric/src/fabric2_server.erl [hunk 6]
```diff
@@ -212,14 +213,47 @@ get_db_and_cluster(EunitDbOpts) ->
...
-            ClusterFileStr = config:get("erlfdb", "cluster_file", ?CLUSTER_FILE),
+            ClusterFileStr = get_cluster_file_string(),
             {ok, ConnectionStr} = file:read_file(ClusterFileStr),
             DbHandle = erlfdb:open(iolist_to_binary(ClusterFileStr)),
             {string:trim(ConnectionStr), DbHandle}
     end,
     apply_tx_options(Db, config:get(?TX_OPTIONS_SECTION)),
     {Cluster, Db}.
 
+get_cluster_file_string() ->
```
Comment (bessbd) [inline src/fabric/src/fabric2_server.erl:224]: Based on the name only I wasn't able to tell if it returns the path or the contents of a cluster file.
How about `get_cluster_file_path` here?

**Review by bessbd — COMMENTED** [review]
Looks like an awesome change. Thank you for this PR, @kocolosk !

File: src/fabric/src/fabric2_server.erl (hunk 1 above)
Comment (rnewson) [reply src/fabric/src/fabric2_server.erl:48]: cc @wohali

Comment (wohali) [reply src/fabric/src/fabric2_server.erl:48]: I wasn't planning on doing WIndows support until all of this stuff settled down... I only want to have to touch it **once**.

Are we really close to a 4.0 release now?

File: src/fabric/src/fabric2_server.erl [hunk 7]
```diff
@@ -212,14 +213,47 @@ get_db_and_cluster(EunitDbOpts) ->
...
+    {error, cluster_file_missing};
+find_cluster_file([{custom, false} | Rest]) ->
+    find_cluster_file(Rest);
+find_cluster_file([{Type, Location} | Rest]) ->
+    case file:read_file_info(Location, [posix]) of
+        {ok, #file_info{access = read_write}} ->
+            couch_log:info("Using ~s FDB cluster file: ~s", [Type, Location]),
+            {ok, Location};
+        {ok, _} ->
+            couch_log:error("CouchDB needs read/write access to FDB cluster file: ~s", [Location]),
```
[the remark this answers is not in the document]
Comment (nickva) [reply src/fabric/src/fabric2_server.erl:248]: Good idea to mention it in erlfdb somewhere referencing https://apple.github.io/foundationdb/administration.html#required-permissions perhaps

File: src/fabric/src/fabric2_server.erl (hunk 1 above)
Comment (kocolosk) [reply src/fabric/src/fabric2_server.erl:48]: I don't think a 4.0 release is imminent, but I can take a shot at extending this PR to do best-effort Windows support.

File: src/fabric/src/fabric2_server.erl (hunk 2 above)
Comment (kocolosk) [reply src/fabric/src/fabric2_server.erl:226]: I had started with that approach. Technically it violates the spec for `os:getenv/2` to use an atom as the second argument, but it does work. Will check if dialyzer complains.

Comment (nickva) [reply src/fabric/src/fabric2_server.erl:226]: Ah makes sense then, if the dializer complains let's keep it as is. Maybe with a short comment about it?

**Comment by kocolosk** [comment]
I would be open to downgrading the permissions check to a non-fatal error message. I suspect we could end up with a lot of situations where CouchDB refuses to start but the ability to modify the cluster file is a non-issue; e.g. it's a single-server deployment, or a cluster where the FDB server nodes are co-located with the CouchDB nodes (and likely able to modify the cluster file themselves).

The situation we're concerned about is where CouchDB is running as a stateless node and only using the local FDB client to communicate with a remote FDB cluster, _and_ that cluster's coordinators end up changing over time, _and_ there's no out-of-band mechanism to ensure the cluster file gets updated on the CouchDB nodes. It's a fairly sophisticated scenario, and I'd be comfortable with CouchDB logging an error message on startup along with some documentation about the appropriate permissions.

File: src/fabric/src/fabric2_server.erl (hunk 7 above)
[the remark this answers is not in the document]
Comment (kocolosk) [reply src/fabric/src/fabric2_server.erl]: New behavior logs this message instead when it finds a read-only cluster file:

> Using read-only default FDB cluster file: /usr/local/etc/foundationdb/fdb.cluster -- if coordinators are changed without updating this file CouchDB may be unable to connect to the FDB cluster!

**Review by nickva — APPROVED** [review]
Retested with unset `FDB_CLUSTER_FILE` (read-only cluster file) and with a writable one. The read-only one emitted the warning in the logs as intended.

Unit tests pass too:
```
======================== EUnit ========================
module 'fabric2_server'
  fabric2_server:380: find_cluster_file_test_ (ignore unspecified config)...[0.069 s] ok
  fabric2_server:388: find_cluster_file_test_ (allow read-only file)...ok
  fabric2_server:395: find_cluster_file_test_ (fail if no access to configured cluster file)...ok
  fabric2_server:402: find_cluster_file_test_ (fail if configured cluster file is missing)...ok
  fabric2_server:410: find_cluster_file_test_ (check multiple default locations)...ok
  [done in 0.084 s]
=======================================================
  All 5 tests passed.
```

Well done!
